public/javascripts/AchievementTracker.js

```diff
@@ -0,0 +1,214 @@
+class AchievementTracker {
+    constructor() {
+        this.mapBadges = {};
+        this.mapBadges[BadgeTypes.Missions] = {
+            1: new Badge(BadgeTypes.Missions, 1, 5),
+            2: new Badge(BadgeTypes.Missions, 2, 10),
+            3: new Badge(BadgeTypes.Missions, 3, 20),
+            4: new Badge(BadgeTypes.Missions, 4, 50),
+            5: new Badge(BadgeTypes.Missions, 5, 100)
+        };
+
+        this.mapBadges[BadgeTypes.Distance] = {
+            // All distances currently in miles.
+            1: new Badge(BadgeTypes.Distance, 1, 0.09), // approximately 500 ft
+            2: new Badge(BadgeTypes.Distance, 2, 0.25),
+            3: new Badge(BadgeTypes.Distance, 3, 0.5),
+            4: new Badge(BadgeTypes.Distance, 4, 1),
+            5: new Badge(BadgeTypes.Distance, 5, 5)
+        };
+
+        this.mapBadges[BadgeTypes.Labels] = {
+            1: new Badge(BadgeTypes.Labels, 1, 50),
+            2: new Badge(BadgeTypes.Labels, 2, 250),
+            3: new Badge(BadgeTypes.Labels, 3, 500),
+            4: new Badge(BadgeTypes.Labels, 4, 1000),
+            5: new Badge(BadgeTypes.Labels, 5, 5000)
+        };
+
+        this.mapBadges[BadgeTypes.Validations] = {
+            1: new Badge(BadgeTypes.Validations, 1, 100),
+            2: new Badge(BadgeTypes.Validations, 2, 250),
+            3: new Badge(BadgeTypes.Validations, 3, 500),
+            4: new Badge(BadgeTypes.Validations, 4, 1000),
+            5: new Badge(BadgeTypes.Validations, 5, 5000)
+        };
+    }
+
+    /**
+     * Gets the current badge given the user's progress for the given badge type.
+     *
+     * @param badgeType: is one of four BadgeTypes: "missions", "distance", "labels", or "validations"
+     * @param value: value corresponds to the current number of completed missions, total distance, num of labels, etc.
+     *               corresponding to the passed badgeType
+     * @returns {null} null if no badge earned or if no badge found for the passed badgeType. Otherwise, the Badge
+     *          corresponding to the highest level earned
+     */
+    getBadgeEarned(badgeType, value) {
+        if (badgeType in this.mapBadges) {
+            let mapLevelsToBadge = this.mapBadges[badgeType]
+            let prevBadge = null;
+            for (let level of Object.keys(mapLevelsToBadge)) {
```

Comment:
I changed your for..in loops to for..of loops, looping over the keys. I think that with the for..in loops you are supposed to do the hasOwnProperty() check that you don't need to do with this other method. I also think that for..in doesn't guarantee ordering, which it looks like you are relying on here.